File: leetcode/trapping_rain_water.cpp
#include <vector>
#include <iostream>

class Solution
{
public:
    static int trap(const std::vector<int> &heights)
    {
        std::vector<int> leftMax(heights.size()), rightMax(heights.size());

        int runningMax{};
        for (size_t i{}; i < leftMax.size(); ++i)
        {
            runningMax = std::max(runningMax, heights[i]);
            leftMax[i] = runningMax;
        }

        runningMax = 0;
        for (size_t i{rightMax.size() - 1}; i != 0; --i)
        {
            runningMax = std::max(runningMax, heights[i]);
            rightMax[i] = runningMax;
        }

        int water{};
        for (size_t i{}; i < heights.size(); ++i)
        {
            water += std::max(0, std::min(leftMax[i], rightMax[i]) - heights[i]);
        }

        return water;
    }
};

int main()
{
    // 6
    std::cout << Solution::trap({0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}) << '\n';

    // 9
    std::cout << Solution::trap({4, 2, 0, 3, 2, 5}) << '\n';
}

File: leetcode/trapping_rain_water.py
class Solution:
    def trap(self, heights: list[int]) -> int:
        if len(heights) < 3:
            return 0

        heights.append(0)

        water = 0
        i = 0
        while i + 1 < len(heights):
            i_height = heights[i]
            # Find one that's higher, or the highest subsequently available otherwise.
            next_big = i + 1
            for j in range(i + 1, len(heights)):
                j_height = heights[j]
                if j_height > heights[next_big]:
                    next_big = j

                if j_height > i_height:
                    break

            pool_height = min(i_height, heights[next_big])
            for between in range(i + 1, next_big):
                water += pool_height - heights[between]

            i = next_big

        return water


if __name__ == "__main__":
    print(Solution().trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))
